### PR TITLE
[ISSUE-79] Unexpected interaction between Union and Optional.

### PR DESCRIPTION
**Pull request by iAndriy**
Address Issue #79

### CODE REVIEW

**Comment by iAndriy** [comment]
@lovasoa Hey Ophir, that PR address use case when group of types supplied to `Union` has None type within it's values:
```
Union[TypeA, TypeB, None]
```

File: marshmallow_dataclass/__init__.py [hunk 1]
```diff
@@ -442,14 +442,16 @@ def field_for_schema(
...
-            metadata["default"] = metadata.get("default", None)
-            metadata["missing"] = metadata.get("missing", None)
-            metadata["required"] = False
-            return field_for_schema(subtyp, metadata=metadata, base_schema=base_schema)
         elif typing_inspect.is_union_type(typ):
+            if typing_inspect.is_optional_type(typ):
+                metadata["default"] = metadata.get("default", None)
+                metadata["missing"] = metadata.get("missing", None)
+                metadata["required"] = False
+            subtypes = [t for t in arguments if t is not NoneType]  # type: ignore
```
Comment (lovasoa) [inline marshmallow_dataclass/__init__.py:450]: Why do we need a `# type: ignore` here ?

**Comment by lovasoa** [comment]
Could you also add the example you mentioned in #79 to the tests ? Trying to deserialize `{'x': {'_type': 'C'}}` with your `A` Schema.

File: marshmallow_dataclass/__init__.py (hunk 1 above)
Comment (iAndriy) [reply marshmallow_dataclass/__init__.py]: thanks, I've removed it.

Comment (iAndriy) [reply marshmallow_dataclass/__init__.py]: @lovasoa Actually it's needed, otherwise mypy hook fails.

**Comment by iAndriy** [comment]
@lovasoa I've added new test case which covers case Optional[Union[A, B]] case.

File: marshmallow_dataclass/__init__.py (hunk 1 above)
Comment (lovasoa) [reply marshmallow_dataclass/__init__.py:450]: So, why is it needed ? What is the type error ?

Comment (iAndriy) [reply marshmallow_dataclass/__init__.py:450]: it's got here as part of modification of ```subtyp = next(t for t in arguments if t is not NoneType)  # type: ignore``` which were in place before. Do you want to dig into that ?

Comment (iAndriy) [reply marshmallow_dataclass/__init__.py:450]: Found pr that line originally came in: https://github.com/lovasoa/marshmallow_dataclass/pull/39/files#diff-be67bc8d0a9f634f273c1261669f88caR392

**Review by lovasoa — APPROVED** [review]

**Comment by CedricCabessa** [comment]
hey, thanks for the fix. It is useful for us!
However, serialization is broken in some case, here is another patch https://github.com/lovasoa/marshmallow_dataclass/pull/107

**Pull request merged**